KEP-4136: Admission Fair Sharing (#4252)
kubernetes-sigs/kueue: keps/4136-admission-fair-sharing/README.md

This model has multiple assumptions:

* Workloads can be preempted.
* Users get a proper quota to do their regular business.
* Users come from multiple teams/organizations and they would rather have a strict policy but 
fair policy than show understanding to those consuming all shared resources. 

However, these assumptions are not universal. Sometimes:

* Workloads should not be preempted.
* Users operate on a shared but bigger quota.
* Fairness should not win over getting the workloads eventually completed.

In that case the existing mode doesn’t work and a different one needs to be employed.

### Goals

* Establish a method for how shared resource usage is calculated and recorded and how users can fine tune the mechanism.
* Allow to specify a fair admission scope at either individual Cluster Queue or Cohort scope.
* Allow to specify the relative importance of LocalQueues targeting the same ClusterQueue.
* Amend the admission mechanism to work on admission scopes instead of only on ClusterQueues. 
* Select the appropriate admission candidates for each of the admission scopes and admit them according to the selected queueing policy.

> * Make the new mechanism complementary to the existing preemption-based fair sharing

Author: Added below.

### Non-Goals

* Store time series data inside K8S.
* Provide precise shared resource usage accounting or billing.

## Proposal

* Modify CQ’s FairSharing struct with 

> type FairSharing struct {

Author: Done.

	// Weight denotes how important the given queue when competing against other queues 
// for unused shared resources. The exact impact of the weight in fair share calculations
// depends on the fair share algorithm used. Default = 1.
	Weight *resource.Quantity `json:"weight,omitempty"`
}

* Expand LQ spec with the same FairSharing struct (Cohort will be expanded with FairSharing 
as a part of the Preemptive FS implementation in hierarchical structure).

* Modify FairSharingStatus struct with 

> type FairSharingStatus struct {

Author: Done.

	// WeightShare represents the usage above nominal quota, with the weight applied in.
	// The bigger the value is the more shared resources has been allocated and the less 
	// entitled the queue is for more shared resources. 
	// The exact details and the interpretation of the value depends on 
    // the fair sharing algorithm used.
	WeightedShare int64 `json:"weightedShare"`

	
	// ConsumedResources represents the aggregated usage of resources over time,
	// with decaying function applied. 
	// The value is populated if usage consumption functionality is enabled in Kueue config.
	ConsumedResources map[string]resource.Quantity `json:"consumedResources,omitempty"`

> 

Reviewer: Do we want to have a field in status for the result of applying weights to `ConsumedResources` so it's easier to monitor and debug?

	// LastUpdate is the time when share and consumed resources were updated.
	LastUpdate metav1.Time `json:"lastUpdate,omitempty"`
}

* Add FairSharingStatus to LocalQueue (and Cohort).

* Create a new struct AdmissionScope and make it an optional field for CQ and Cohort Spec. If
not provided, CQ or Cohort is not considered an AdmissionScope and is not a subject for new

> admission logic. If there are two AdmissionScopes on the path from CQ/Cohort to the top of

Author: Because someone may be changing the scope. Scope change is not atomic and we cannot block the entire hierarchy in the meantime.

the hierarchy tree, the higher one is used.

Const (
	// FairSharing based on usage, with QueuingStrategy as defined in CQ.
	UsageBasedFairSharing AdmissionMode = UsageBasedFairSharing 

> NoFairSharing AdmissionMode = NoFairSharing	

Reviewer: Does it differ from excluding a CQ/Cohort from the new logic?

Author: This is exactly this. However CQ may be in a cohort that is an admission scope. Marking CQ as no fair sharing will skip any fair sharing among LQs.

)

type AdmissionScope struct {
AdmissionMode AdmissionMode
}

* When selecting candidates for admission groups all workloads from LQ to CQ or the
topmost Cohort that is marked with AdmissionScope. Then sort them using criterias:

> - Usage vector  built from ConsumedResources from TopCohort to LQ

Reviewer: What metric should we use compare two vectors? Is there any dominant resource we should pay the most attention to?

Reviewer: Okay, later you mention `resourceWeights` API in Kueue config

Author: Yes.

- Priority
- Timestamp

Let’s look at a couple scenarios:

1. AdmissionScope at CQ, CQ queueing policy is FIFO, 3 LQ pointing to CQ. Kueue considers
all CQ resources and potentially borrowed resources as “shared” resources and fair sharing
is applied to all workloads.

> Kueue sorts the workloads by their LQ usage (if mode is not NoFairSharing), priority and

Reviewer: ```suggestion
Kueue sorts the workloads by their LQ usage (if mode is `UsageBasedFairSharing`), priority and
```

Author: Done

timestamp and tries to admit the first one from the list. Other workloads are not attempted
until the first one is not admitted.

2.  AdmissionScope at CQ, CQ queueing policy is BestEffort, 3 LQ pointing to CQ. 
Kueue considers all CQ resources and potentially borrowed resources as “shared” resources and
fair sharing is applied to all workloads.

> Kueue sorts the workloads by their LQ usage (if mode is not NoFairSharing), priority and

Reviewer: ```suggestion
Kueue sorts the workloads by their LQ usage (if mode is`UsageBasedFairSharing`), priority and
```

Author: Done.

timestamp and tries to admit the first one from the list. If it fails and the second, third

> or following is possible then that workload is admitted, under condition that it might get preempted.

Author: In best-effort queues workloads may hop over the first one, under condition that the first workload may preempt them if the sum of free resources and these consumed by skip-the-line workloads is greater than the needs of the first workloads. 

We may not support this in the first iteration but KEP should at least briefly explain what it could look like if we decided to complete the picture.

3. AdmissionScope at Cohort level - Kueue operates in a mixed mode. Inside CQ workloads are

> selected according to their AdmissionMode (if specified). If a workload fits entirely into

Author: The candidates inside individual CQs are selected based on the specified logic and bubbled up.

nominal quota, then it is admitted immediately, if not it goes into cohort-level fair sharing.

> For Cohort we select all the “sticking out” workloads, and sort them by their CQ usage, priority

Author: Updated/

and timestamp. Kueue attempts to admit the first workload from the list of sticking-out, just

> like if it was one big strict FIFO queue. For multi-level hierarchy under one AdmissionScope

Author: Probably the second but I guess it is a bit of an implementation detail. I would be ok with both.

we would treat underlying Cohorts as Fifo CQs.

* Additionally there will be the resource usage calculation loop. The frequency of the
calculation will be controlled globally in Kueue’s config file. Accounting would be done
using something like geometric average:

usage_sum = (1-A) * previous_usage_sum + A * current_usage.  

> The value will be stored in FairSharingStatus for all LQ, CQ, and Cohorts.

Author: Where would you put it otherwise?

> The value will be stored in FairSharingStatus for all LQ, CQ, and Cohorts.

Author: Yes, the value stays after brief downtime but is restarted after a longer one. Updated.

The user will be able to configure the decaying factor A in Kueue’s config file by specifying
the half decay time - after what time the current shared usage will decay to half of its original value.

A = 1 - 0.5 ^ (sampling/half_decay)

* Configuration will sit in FairSharing stuct in Kueue config. There will be the following modifications:

  - modes - decides which modes are enabled in FairSharing is turned on. By default it would be: [“PreemptionBased”]. Allowed values: “PreemptionBase”, “AdmissionTime”.
  - usageHalfDecayTime - half decay time of usage, as described above.
  - usageSamplingFrequency - how often usage is calculated.

>   - resourceWeights - how much consumption of individual resources is important when comparing usage.

Reviewer: Could you expand on this? Are there any defaults for this?

## Motivation

Currently Kueue has a Fair Sharing mechanism that enforces fair-sharing of unused resources
via preemption. If one Cluster Queue is using much more of the resources than the other one
that is in need, some workloads from the first one may be preempted to allow more “fair” distribution. 

This model has multiple assumptions:

* Workloads can be preempted.
* Users get a proper quota to do their regular business.
* Users come from multiple teams/organizations and they would rather have a strict policy but 
fair policy than show understanding to those consuming all shared resources. 

However, these assumptions are not universal. Sometimes:

* Workloads should not be preempted.
* Users operate on a shared but bigger quota.
* Fairness should not win over getting the workloads eventually completed.

In that case the existing mode doesn’t work and a different one needs to be employed.

### Goals

* Establish a method for how shared resource usage is calculated and recorded and how users can fine tune the mechanism.
* Allow to specify a fair admission scope at either individual Cluster Queue or Cohort scope.

> * Allow to specify the relative importance of LocalQueues targeting the same ClusterQueue.

Reviewer 2: IMO this is solved via hierarchical cohorts and the `fairSharing.weight` I think we shouldn't complicate the LocalQueue API. Maintaining it as a pointer to a ClusterQueue feels the most natural to me rather than splitting configuration details

Author: Yes, we could approximate the solution by making admission scope at the cohort level only and use cluster queues as entry points/proxies/local queues. However I'm not 100% sure that we need this level of complication. The KEP covers it just in case so we have the full picture, but suggests starting small. Maybe doing admission fair sharing only at CQ+LQ level would be enough for most users needing the feature. Implementation-wise CQ+LQ is much simpler and introduces far less changes to already complex admission algorithm than going full hierarchical.

* Amend the admission mechanism to work on admission scopes instead of only on ClusterQueues. 
* Select the appropriate admission candidates for each of the admission scopes and admit them according to the selected queueing policy.
* Make the new mechanism complementary to the existing preemption-based fair sharing

### Non-Goals

* Store time series data inside K8S.
* Provide precise shared resource usage accounting or billing.

> ## Proposal

Reviewer 3: Could you expand a bit more and clarify the status of preemption and reclamation. In particular, if the new fair sharing mode supports priority based preemption, or reclamation? So, if it supports reclamation, how are we going to select the candidates: as in the "classical preemption", the "classical fair sharing" , or would there be a new algo?

Reviewer 3: So even both fair sharing approaches can work at the same time for a cohort, ie. The new fair sharing will order workloads for scheduling, and the old fair sharing will determine reclamation targets. Right?

Reviewer 3: This makes sense, thanks for clarifying. 

Still, the clash of naming is confusing, as these are not just two modes of the same.

So I'm thinking about renaming the feature to avoid the collision. I guess the main "selling point" about the feature is being past usage aware, and making "fair" admission choices in that regard. 
1. Fair Admission
2. Past-Aware Admission
3. Past Usage-Aware Admission
4. Past Usage-Aware Fair Sharing

I'm leaning to (3.) to emphasize that the "past usage" awareness is the main feature here. wdyt?

cc @gabesaba @PBundyra @pajakd if you have some other thoughts.

Reviewer: Not sure how we should proceed with this, but I'm not the biggest fan of how extensive 2,3,4 are. Also not sure whether the name clash is a bad thing - we reflect those two modes also in API, both in config, and in objects' status

Reviewer 3: >  I'm not the biggest fan of how extensive 2,3,4 are

You mean long? I think `Past Usage-Aware Admission` (my fav) isn't really much longer than others. 

> we reflect those two modes also in API, both in config, and in objects' status

Yeah, and I find this is confusing too. This is because this feature is not a new mode of "Fair Sharing" (as the name suggests). It can actually be used in combination with the current Fair Sharing as pointed out in https://github.com/kubernetes-sigs/kueue/pull/4252#discussion_r2045007906.

Reviewer 3: Actually, after syncing with @mwielgus and @PBundyra I'm ok with the current naming. The argument is that the new feature is conceptually a new mode of fair sharing at the level of admission. The "Past Usage-Awareness" is an important aspect of the current implementation, but maybe another approach will be used in the future. 

Still, I believe the naming can be confusing here, but we can address it at the documentation level. Feel free to resolve.

```go
type FairSharing struct {
	// Weight denotes how important the given queue when competing against other queues 
// for unused shared resources. The exact impact of the weight in fair share calculations
// depends on the fair share algorithm used. Default = 1.
	Weight *resource.Quantity `json:"weight,omitempty"`
}
```

* Expand LQ spec with the same FairSharing struct (Cohort will be expanded with FairSharing 
as a part of the Preemptive FS implementation in hierarchical structure).

* Modify FairSharingStatus struct with 

```go
type FairSharingStatus struct {
	// WeightShare represents the usage above nominal quota, with the weight applied in.
	// The bigger the value is the more shared resources has been allocated and the less 
	// entitled the queue is for more shared resources. 
	// The exact details and the interpretation of the value depends on 
    // the fair sharing algorithm used.
	WeightedShare int64 `json:"weightedShare"`

	
	// ConsumedResources represents the aggregated usage of resources over time,
	// with decaying function applied. 
	// The value is populated if usage consumption functionality is enabled in Kueue config.

> 	ConsumedResources map[string]resource.Quantity `json:"consumedResources,omitempty"`

Reviewer: ```suggestion
	ConsumedResources map[corev1.ResourceName]resource.Quantity `json:"consumedResources,omitempty"`
```

	// LastUpdate is the time when share and consumed resources were updated.
	LastUpdate metav1.Time `json:"lastUpdate,omitempty"`
}
```

* Add FairSharingStatus to LocalQueue (and Cohort).

> * Create a new struct AdmissionScope and make it an optional field for CQ and Cohort Spec. If

Reviewer 3: How is actually the new fair sharing mode enabled? IIUC this is by the presence of the AdmissionMode on the LQ/CQ or Cohort. However, the "classical fair sharing" is enabled globally, so can we mix them? 

In particular, what if "classical fair sharing" is enabled globally, but the new mode only in LQ, then the rest of CQ and the Cohort would use the classical fair sharing? I'm wondering this can be challenging to implement, and difficult for users to understand the configuration options.

Author: Yes, we can mix the. Added more description around line 180,

not provided, CQ or Cohort is not considered an AdmissionScope and is not a subject for new
admission logic. If there are two AdmissionScopes on the path from CQ/Cohort to the top of
the hierarchy tree, the higher one is used.

```go
const (
	// FairSharing based on usage, with QueuingStrategy as defined in CQ.
	UsageBasedFairSharing AdmissionMode = UsageBasedFairSharing 

> NoFairSharing AdmissionMode = NoFairSharing	

Reviewer 3: Align the enum.

> NoFairSharing AdmissionMode = NoFairSharing	

Reviewer 3: What does "NoFairSharing" actually mean - no new mode of fair sharing, or disable any fair sharing - also "classical"?

selected according to their AdmissionMode (if specified). If a workload fits entirely into
nominal quota, then it is admitted immediately, if not it goes into cohort-level fair sharing.
For Cohort we select all the “sticking out” workloads, and sort them by their CQ usage, priority
and timestamp. Kueue attempts to admit the first workload from the list of sticking-out
(workload + current_usage > resources), just like if it was one big strict FIFO queue. 
For multi-level hierarchy under one AdmissionScope we would treat underlying Cohorts as Fifo CQs.

* Additionally there will be the resource usage calculation loop. The frequency of the
calculation will be controlled globally in Kueue’s config file. Accounting would be done
using something like geometric average:

usage_sum = (1-A) * previous_usage_sum + A * current_usage.  

The value will be stored in FairSharingStatus for all LQ, CQ, and Cohorts.

The user will be able to configure the decaying factor A in Kueue’s config file by specifying
the half decay time - after what time the current shared usage will decay to half of its original value.

> A = 1 - 0.5 ^ (sampling/half_decay)

Reviewer: ```suggestion
A = 1 - 0.5 ^ (half_decay/sampling)
```
Can we do it like this so the result of dividing is likely to be >1 instead of <1

* Configuration will sit in FairSharing stuct in Kueue config. There will be the following modifications:

> 

Reviewer: Can we agree on more concrete shape of the config API, as the follow-up to this discussion: https://github.com/kubernetes-sigs/kueue/pull/4687#discussion_r2013886300

Author: What would be your proposal?

Reviewer: I'd propose to embed `usageHalfLifeDecayTime`, `usageSamplingInterval`, `resourceWeights` into a struct called `AdmissionFairSharing` and make it a pointer. Then the presence of non-nil value of this pointer could determine if the Admission mode is switch on/off, so there's no possibility of having invalid state of the configuration. 

Ideally, in the next version of API I think we should drop the `Enable` field and migrate towards a struct `PreemptionFairSharing` which presence would determine if it's switched on/off

Author: Ok.

Reviewer: Since you agree can we remove `modes` field?

Author: Done.

  - modes - decides which modes are enabled in FairSharing is turned on. By default it would be: [“PreemptionBased”]. Allowed values: “PreemptionBase”, “AdmissionTime”.
  - usageHalfDecayTime - half decay time of usage, as described above.

>   - usageSamplingFrequency - how often usage is calculated.

Reviewer 4: "HalfDecayTime" -- If this is inspired by natural sciences then half-life is the term we might want to use.
'SamplingFrequency' -- (also in natural sciences 🤓) frequency is the number of occurrences of sth in a unit of time. If here this parameter will define the time between refreshes of the values then this is rather sampling interval or sampling period

Author: OK, updated.

* Increased complexity of the project.

## Design Details

Covered in Proposal.

### Test Plan
[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.
##### Prerequisite testing updates

#### Unit Tests
The code will be thoroughly covered with unit tests.

In particular:
* CQ level fair sharing between LQ for both strict FIFO and best effort.
* Cohort with owned resources, and CQ with guaranteed quota.

#### Integration tests

They will mainly focus on larger scope scheduling (involving multiple cohorts/cqs) and
interactions with preemptive fair sharing outside admission scope.

### Graduation Criteria

The implementation will be split into 2 subfeatures:

* CQ + LQ level support
* Multi-level Cohort+CQ+LQ support

Obviously, the second depends on the first to some extent. The first however may reach
Beta/GA without starting the second.

The graduation criterias are quite standard:

* Beta - positive feedback from Alpha, api seems reasonable.
* GA - positive feedback, no bugs, no api changes needed.

> We hope to have CQ+LQ in alpha for the next Kueue release (0.11).

Reviewer: ```suggestion
We hope to have CQ+LQ in alpha for the next Kueue release (0.12).
```

### Goals

* Establish a method for how shared resource usage is calculated and recorded and how users can fine tune the mechanism.
* Allow to specify a fair admission scope at either individual Cluster Queue or Cohort scope.
* Allow to specify the relative importance of LocalQueues targeting the same ClusterQueue.
* Amend the admission mechanism to work on admission scopes instead of only on ClusterQueues. 
* Select the appropriate admission candidates for each of the admission scopes and admit them according to the selected queueing policy.
* Make the new mechanism complementary to the existing preemption-based fair sharing

### Non-Goals

* Store time series data inside K8S.
* Provide precise shared resource usage accounting or billing.

## Proposal

* Modify CQ’s FairSharing struct with 

```go
type FairSharing struct {
	// Weight denotes how important the given queue when competing against other queues 
// for unused shared resources. The exact impact of the weight in fair share calculations
// depends on the fair share algorithm used. Default = 1.
	Weight *resource.Quantity `json:"weight,omitempty"`
}

> ```

Reviewer 5: IIUC, we already have this field

https://github.com/kubernetes-sigs/kueue/blob/636d57aca1d4500aa6a09c240a8d0cc341f0e4f2/apis/kueue/v1beta1/clusterqueue_types.go#L128-L132

https://github.com/kubernetes-sigs/kueue/blob/636d57aca1d4500aa6a09c240a8d0cc341f0e4f2/apis/kueue/v1beta1/fairsharing_types.go#L39

Author: Yes, there is. I change the meaning a bit to be more generic.

Reviewer 5: I see. We will change only the meaning, and not change the API's looking. Thanks.

* Expand LQ spec with the same FairSharing struct (Cohort will be expanded with FairSharing 
as a part of the Preemptive FS implementation in hierarchical structure).

* Modify FairSharingStatus struct with 

```go
type FairSharingStatus struct {

> 	// WeightShare represents the usage above nominal quota, with the weight applied in.

Reviewer 5: ```suggestion
	// WeightedShare represents the usage above nominal quota, with the weight applied in.
```

	// The bigger the value is the more shared resources has been allocated and the less 
	// entitled the queue is for more shared resources. 
	// The exact details and the interpretation of the value depends on 
    // the fair sharing algorithm used.
	WeightedShare int64 `json:"weightedShare"`

	
	// ConsumedResources represents the aggregated usage of resources over time,
	// with decaying function applied. 
	// The value is populated if usage consumption functionality is enabled in Kueue config.

> 	ConsumedResources map[corev1.ResourceName]resource.Quantity `json:"consumedResources,omitempty"`

Reviewer 5: ```suggestion
	ConsumedResources corev1.ResourceList `json:"consumedResources,omitempty"`
```

Why not `corev1.ResourceList`?

	// LastUpdate is the time when share and consumed resources were updated.

> 	LastUpdate metav1.Time `json:"lastUpdate,omitempty"`

Reviewer 5: Why not use conditions? 
In this approach, I think If we want to add any reason and message for FairShare, we need to add those condition similarity fields here.

Author: Conditions have last transition, not last update. 
```
// lastTransitionTime is the last time the condition transitioned from one status to another.
// This should be when the underlying condition changed. 
```
Here there is no transition, just updates.

Reviewer 5: I see. You mean this is the time when kueue snapshot consumed usage.
In that case, we probably want to say `usageSnapshotAt`, `usageCalculationAt` something so that we can obviously clarify what is this time.

WDYT?

Author: Personally I see `LastUpdate` time as more straightforward and potentially covering also `WeightedShare` or whatever else we add in the future. UsageSnapshotAt limits us, and if any new field is added then we will need another timestamp.

	// entitled the queue is for more shared resources. 
	// The exact details and the interpretation of the value depends on 
    // the fair sharing algorithm used.
	WeightedShare int64 `json:"weightedShare"`

	
	// ConsumedResources represents the aggregated usage of resources over time,
	// with decaying function applied. 
	// The value is populated if usage consumption functionality is enabled in Kueue config.
	ConsumedResources map[corev1.ResourceName]resource.Quantity `json:"consumedResources,omitempty"`

	// LastUpdate is the time when share and consumed resources were updated.
	LastUpdate metav1.Time `json:"lastUpdate,omitempty"`
}
```

* Add FairSharingStatus to LocalQueue (and Cohort).

* Create a new struct AdmissionScope and make it an optional field for CQ and Cohort Spec. If
not provided, CQ or Cohort is not considered an AdmissionScope and is not a subject for new
admission logic. If there are two AdmissionScopes on the path from CQ/Cohort to the top of
the hierarchy tree, the higher one is used.

```go
const (
	// FairSharing based on usage, with QueuingStrategy as defined in CQ.
	UsageBasedAdmissionFairSharing AdmissionMode = UsageBasedFairSharing

>     NoAdmissionFairSharing AdmissionMode = NoAdmissionFairSharing

Reviewer 3: Maybe just "DefaultAdmission"?

Author: Default doesn't say what it is doing. Some strategy may nominated as default but should not be named as such.

Reviewer 3: ok, feel free to resolve

usage_sum = (1-A) * previous_usage_sum + A * current_usage.  

The value will be stored in FairSharingStatus for all LQ, CQ, and Cohorts. The value will not be zeroed
after Kueue restart or after brief period of downtime. However if the period is longer,
the value should be automatically zeroed.

The user will be able to configure the decaying factor A in Kueue’s config file by specifying
the half life decay time - after what time the current shared usage will decay to half of its original value.

A = 1 - 0.5 ^ (sampling/half_life_decay)

* Configuration will sit in FairSharing stuct in Kueue config. There will be the following modifications:

  - modes - decides which modes are enabled in FairSharing is turned on. By default it would be: [“PreemptionBased”]. Allowed values: “PreemptionBase”, “AdmissionTime”.
  - usageHalfLifeDecayTime - half life decay time of usage, as described above.
  - usageSamplingInterval - how often usage is calculated.
  - resourceWeights - how much consumption of individual resources is important when comparing usage.
  - resetInactivityPeriod - if Kueue has not updated the value for this period then the value should be zeroed.
 

> If the user doesn't want any preemptions while fair sharing, preemptionStrategies should be left empty.

Reviewer: Does it mean `.enable` field is redundant, as PreemptionStrategies determine if preemption fs should be applied?

Author: I would say that enable turns all FS on/off.

Reviewer: What happens if users sets `enable=True` but leaves `preemptionStrategies` empty as well as `.admissionFairSharing` field

Author: They may get admission fair sharing, Kueue will calculate preemption-based scores but no fair-sharing preemption will be triggered.
